token/core/fabtoken/validator.go

```diff
@@ -9,24 +9,95 @@ package fabtoken
 import (
 	"bytes"
 	"encoding/json"
+	"time"
 
 	"github.com/hyperledger-labs/fabric-smart-client/platform/view/services/hash"
 	"github.com/hyperledger-labs/fabric-smart-client/platform/view/view"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/core/identity"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/core/interop"
 	"github.com/hyperledger-labs/fabric-token-sdk/token/driver"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/interop/exchange"
 	token2 "github.com/hyperledger-labs/fabric-token-sdk/token/token"
 	"github.com/pkg/errors"
 	"go.uber.org/zap/zapcore"
 )
 
+var defaultValidators = []ValidateTransfer{SerializedIdentityTypeExtraValidator, ScriptTypeExchangeExtraValidator}
+
+type ValidateTransfer func(inputTokens []*token2.Token, tr driver.TransferAction) error
+
+func SerializedIdentityTypeExtraValidator(inputTokens []*token2.Token, tr driver.TransferAction) error {
+	// noting else to validate
+	return nil
+}
+
+func ScriptTypeExchangeExtraValidator(inputTokens []*token2.Token, tr driver.TransferAction) error {
+	for _, in := range inputTokens {
+		owner, err := identity.UnmarshallRawOwner(in.Owner.Raw)
+		if err != nil {
+			return errors.Wrap(err, "failed to unmarshal owner of input token")
+		}
+		if owner.Type == exchange.ScriptTypeExchange {
+			if len(inputTokens) != 1 || len(tr.GetOutputs()) != 1 {
+				return errors.Errorf("invalid transfer action: an exchange script only transfers the ownership of a token")
+			}
+
+			out := tr.GetOutputs()[0].(*TransferOutput).Output
+			if inputTokens[0].Type != out.Type {
+				return errors.Errorf("invalid transfer action: type of input does not match type of output")
+			}
+			if inputTokens[0].Quantity != out.Quantity {
+				return errors.Errorf("invalid transfer action: quantity of input does not match quantity of output")
+			}
+
+			// check that owner field in output is correct
+			if err := interop.VerifyTransferFromExchangeScript(inputTokens[0].Owner.Raw, out.Owner.Raw); err != nil {
+				return errors.Wrap(err, "failed to verify transfer from exchange script")
+			}
+		}
+	}
+
+	for _, o := range tr.GetOutputs() {
```

Comment:
it is not clear to me why we check here for redeem ? this check should also be peformed when there is not exchange script validator.

Reply:
We check here just to make sure we don't create an error when trying to unmarshall later
There was no check like this performed by the non-script case